Initial support for GC ref map emission in the CPAOT compiler

trylek opened the pull request:
I have basically forked away ArgIterator and TransitionBlock from
the calling convention converter and I adapted it to use by the
CPAOT compiler. I have not yet added multi-architecture support,
for now it only works for the compiler build architecture. There are
also a few NotImplemented cases I'll fix once I hit them in testing.

My initial thinking is that, after finalizing this code and after
we start working on other architectures, I'll further refactor this
code based on JanK's suggestion by merging in preprocessed versions
of the ArgIterator source file for the four architectures.

I also plan to further experiment with runtime architecture
virtualization that I tried to prototype in a trimmed-down
clone of TransitionBlock I coded up for the purpose of R2RDump
GC ref map decoder and I'm quite pleased with the result which
looks quite elegant and compact to me.

Thanks

Tomas

trylek commented:
OK, I believe that at least anecdotally I see identical GC ref map info in CPAOT vs. Crossgen-produced R2R excutables. Please let me know what you'd like me to change before merging this in; I'd love to proceed with this change as fast as I can because arbitrary tests are randomly failing with this and make debugging annoying.

My current thinking is that, as next step, I'll work on making the change cross-architecture, either by merging in the four preprocessed versions of the transition block and arg iterator source files as JanK originally suggested, or by switching over all architecture differences to runtime conditional checks as I prototyped in my R2RDump GC ref map parsing changes.

jkotas commented on src/ILCompiler.ReadyToRun/src/Compiler/DependencyAnalysis/ReadyToRun/GCRefMapNode.cs:
```diff
@@ -0,0 +1,121 @@
+                IMethodNode methodNode = _methods[methodIndex];
+                if (methodNode is LocalMethodImport localMethodImport && localMethodImport.MethodCodeNode.IsEmpty)
+                {
+                    // Emit an empty entry into the GC ref map for the uncompilable method
```
Is this correct? I would think we need the GCRef even for methods that are not compiled into the image.

trylek replied:
Well, I think the purpose of this conditional block is kind of the other way around - this is to cater for uncompilable methods (e.g. because they reference S.N.Vector) by emitting an empty GC ref map just so that the method import cell array and the GC ref map array don't fall out of sync.

jkotas replied:
I think we need the GC ref map even for methods that we were not able to compile. Otherwise, we won't be able to do GC stack scan when those methods are called at runtime and we fallback to JIT.

trylek replied:
Hmm, how can we encode GC ref map for a function with a Vector parameter if Vector has runtime-determined size?

jkotas replied:
There are two options how to deal with variable sizes Vector:
- Do not compile any methods whose code depends on the variable-sized Vector. It is what crossgen does today.
- Compile methods whose depend on Vector size and emit the `READYTORUN_FIXUP_Check_TypeLayout` fixup that will encode the Vector size assumed by the method code. This fixup will cause the method code to be rejected at runtime if the runtime Vector size is different from the Vector size assumed by the method code.
   - Advanced version of this is the code multi-versioning that we keep talking about every once in a while.

trylek replied:
Thanks for the clarification, this makes sense to me. You have also nicely ordered the three options by their complexity. In this stage I'm heads down pulling off at least option (1); once I add support for type layout checks, I can implement (2) and we can start a technical discussion about (3).

I assume that this most complex option will likely require some additions to the R2R format as I suppose we should end up with a map of variant entrypoints keyed by some processor extension identifier. I don't yet have a full picture of all the extensions / coprocessor / instruction set extensions that are available in practice and supported by JIT; whether they are all mutually exclusive or whether there are some "orthogonal" sets of extensions that may or may not be optionally present on a target machine (so that we'd need a bitmask for the extension ID).

trylek replied:
In fact I'm just an idiot, I just had a blind spot and you were right in the very first place - when a method fails compiling (e.g. because it uses runtime JIT), we still need to be able to emit GC ref map for calls to it. I have hit this in one of the Top200 tests after I started running CPAOT tests with GC stress per your suggestion. I already have a local fix I'm testing right now.

trylek commented:
@MichalStrehovsky For checking in the relevant bits of CoreCLR source code, I think that is a good idea. Sadly the GC ref map logic is scattered over a larger number of files - corcompile.h, gcrefmap.h, compile.cpp, compile.h, zapimport.h, callingconvention.h, cgencpu.h for the four architectures - what is the established practice? Check in all source code that is related or synthesize a new source file from the various bits in the multiple files? In compile.h, for instance, the only relevant bit is the declaration of the virtual method GetCallRefMap.

jkotas commented:
I guess Michal had just the calling convention headers in mind.

trylek commented:
@jkotas Hmm, funnily enough the calling convention is the only part I didn't actually port from CoreCLR, I used the managed version in CoreRT, I just fixed the transition block layouts to CoreCLR format based on JanV's suggestion :-). I even believe the CoreCLR source code for the CCC is already checked in the CoreRT repo.

jkotas commented:
I know, but it is not easy to follow:

Both ArgIterator.cs https://github.com/dotnet/corert/pull/6681/files#diff-4936812eb1717467e7b16e25464bb415R7 and TransitionBlock.cs https://github.com/dotnet/corert/pull/6681/files#diff-834fe8c908a4ad457929bb3b77db8c86R6 have the same comment "This file is a line by line port of callingconvention.h from the CLR"

This comment is not actually correct anymore because of there is a bunch of the CallingConverter stuff mixed in. Also, it is not obvious why line-by-line of same file results into two completely different files...

trylek commented:
Fair enough. The comment is definitely not accurate. Probably the important point is that this is not primarily about the comment, it's rather about figuring out how to make the code easy to maintain. I basically took the CCC code and pruned it to remove unneeded stuff; subsequently I audited TransitionBlock based on the cgencpu files.

These files would be certainly useful to check in as they contain the actual lists of registers and other details affecting transition block layout and GC ref map calculation. For the transition block and arg iterator as such, I think it might be too limiting to maintain them as straight ports from CoreCLR because it wouldn't let us improve them in any manner.

For instance, in my recent R2RDump work on parsing the GC ref map, I experimented with dynamic target architecture selection at runtime and the result was far simpler and more elegant than I expected. I would love to try to do the same experiment with the entire arg iterator - I have a hunch that, with proper virtualization, it should be relatively easy to get rid of all the conditional compilation without making the code more complex, rather making it much simpler and easier to reason about.

trylek commented on src/ILCompiler.ReadyToRun/src/Compiler/DependencyAnalysis/ReadyToRun/GCRefMapBuilder.cs:
```diff
@@ -0,0 +1,394 @@
+                throw new NotImplementedException();
+            }
+
+            if (type is DefType defType)
```
Good suggestion, fixed in 4th commit.

trylek commented on src/ILCompiler.ReadyToRun/src/Compiler/DependencyAnalysis/ReadyToRun/GCRefMapBuilder.cs:
```diff
@@ -0,0 +1,394 @@
+            if (type is DefType defType)
+            {
+                FieldLayoutAlgorithm fieldLayoutAlgorithm = _factory.TypeSystemContext.GetLayoutAlgorithmForType(defType);
+                ComputedInstanceFieldLayout instanceFieldLayout = fieldLayoutAlgorithm.ComputeInstanceLayout(defType, InstanceLayoutKind.TypeAndFields);
```
Good point, fixed in 4th commit.

trylek commented:
In 7th commit I have fixed one remaining deficiency of my initial GC ref map implementation - emitting GC ref map entries for method-like helper / precode cells. I have also rebased the change against the latest master. With this change I have received several successive Top200 CPAOT runs with 0 failures.

@JKotas - can you please advise me how to locally test the change in GC stress mode? IIRC you mentioned before that GC stress should substantially increase the likelihood of runtime failures due to incorrect and / or incomplete GC ref map info so I'm hoping that such testing should increase our confidence in the change. I see that the script

<pre>
tests\CoreCLR\runtest\runtest.cmd
</pre>

supports a "gcstresslevel" parameter - what value should I be using? Is this flag sufficient or do I need to make additional private changes to ILC or to the build / test scripts to fully enable GC stress testing?

Thanks a lot

Tomas

jkotas commented:
> what value should I be using?

You want to stress managed/unmanaged transitions. `3` should give you what you need.

trylek commented:
FYI I have amended the latest commit as I missed a rebase fix in the original commit.

trylek commented:
@jkotas & @MichalStrehovsky - could you please share your thoughts towards these changes in general? I tried to address some of your early feedback and I'm relatively satisfied with the CQ of these changes in terms of CPAOT pass rate in GC stress runs but the changes constitute a broad refactoring of the CoreCLR code; I'm trying to keep improving on the changes but I'd certainly love to hear as fast as I can if either of you is fundamentally opposed to the direction I'm trying to pursue.

Thanks a lot

Tomas

trylek commented:
Thanks Jan for your supportive feedback. I wasn't expecting you to scrutinize handling every possible parameter combination on all architectures, it is certainly on my plate to continue fixing the change as necessary, I was just trying to make sure I'm not steering to some territory you're fundamentally uncomfortable with.

In general, I was originally somewhat uncertain but I'm starting to really like the modified code. Its obvious downside is that it naturally starts to diverge from the CoreCLR "source", however I believe it to bring new benefits that outweigh the cost - apart from not having to check in the preprocessed versions of the source code I believe the code is gradually becoming easier to read; we're now also building code for all architectures every time so it's easier to spot and fix bugs that affect only one of these.
